File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int noZeroProduct=1;
        int zeroCount=0;
        for(int n:nums){
            if(n==0){
                zeroCount++;
            }else{
                noZeroProduct *=n;
            }
        }
        vector<int>ans(nums.size());
        int i=0;
        for(int n:nums){
            if(n==0 && zeroCount==1){
                ans[i]=noZeroProduct;
            }else if(n!=0 && zeroCount>0){
                ans[i]=0;
            }else if(n!=0 && zeroCount==0){
                ans[i]=(noZeroProduct/n);
            }
            i++;
        }
        return ans;
    }

};

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java


class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        if(n==1) return nums;
        int[] leftProd = new int[n];
        leftProd[0] = nums[0];
        for(int i=1;i<n;i++){
            leftProd[i] = leftProd[i-1] * nums[i];
        }
        int rightProd = nums[n-1];
        leftProd[n-1] = leftProd[n-2];
        for(int i=n-2;i>0;i--){
            leftProd[i] = leftProd[i-1] * rightProd;
            rightProd *= nums[i];
        }
        leftProd[0] = rightProd;
        return leftProd;
    }
}